util.lua

```diff
@@ -6,20 +6,47 @@ local function get_single_string(item)
 	return item:to_string()
 end
 
+local craft_result_cache = {}
+
+function util.invalidate_craft_result_cache(pos)
+	craft_result_cache[minetest.pos_to_string(pos)] = nil
+end
+
+function util.get_craft_result(pos)
+	-- cache the craft result on Sokomine's recommendation
```

Comment:
Insufficient function description. "What does it do? Why do we need it? Is there an actual measurable performance difference?" Things like that.

Reply:
i'm not actually sure if this has a measurable performance effect, i was just recruited to deal w/ the following output from someone else's profiler:

| instrumentation                                         |    min µs |    max µs |    avg µs | min % | max % | avg % |
| -------- | -------- | -------- | -------- | -------- | -------- | -------- |
| crafting_bench | 0 | 733629 | 1132 | 0.0 | 99.3 | 0.3 |
|ABM[1] | 2 | 733629 | 103996 | 0.0 | 99.3 | 30.7 |

mentioning Sokomine is not appropriate though, i'll remove that. 

note that this is from a server w/ nearly 64000 registered recipes, i'm not sure how efficient `minetest.get_craft_result` is, though it looks like it's doing a finite number of hash lookups, so perhaps this is just added complexity.

Reply:
i've not been successful in creating much lag w/ the old code in my test environment, and because timers are not profiled, i've got no real way to measure the performance difference.

Reply:
If `minetest.get_craft_result` does scale badly, then it should be reported to minetest/minetest instead of "fixing" it in a single mod.

Reply:
i've done some testing. 

```lua
local recipes = {}
for item in pairs(minetest.registered_items) do
 table.insert_all(recipes, minetest.get_all_craft_recipes(item) or {})
end

local start = os.clock()

for i = 1, #recipes do
 minetest.get_craft_result(recipes[i])
end

local elapsed = os.clock() - start
print(string.format("%.01f us per call", 1e6 * elapsed / #recipes))
```

when running just minetest_game locally, this reports about 9.6us/call. when running a full server test environment locally, this reports 533us/call. when running on the actual game server but under a debug build, this reports 5914us/call, which is a significant fraction of a server step (44ms). if you think this needs to be reported against the engine, i'll do that, but in the meantime, caching the results certainly *looks* useful for our case.